### PR TITLE
fix(50750): Object type literal with string literal property in contextual typing position causes language service error on all literal type references

### LINKED ISSUE #50750: Object type literal with string literal property in contextual typing position causes language service error on all literal type references
**In a TypeScript file**

```ts
interface SomeType {
    property: "foo";
}

/**
 * @type {{ property: "foo"}}
 */
const obj: SomeType = {
    property: "foo",
};
```

Request find-all-references on any string literal. This currently produces the following language service error:

```
<semantic> Response received: references (1602). Request took 7 ms. Success: false . Message: Error processing request. Cannot read properties of undefined (reading 'flags')
TypeError: Cannot read properties of undefined (reading 'flags')
    at getMembersOfSymbol (c:/Users/Daniel/.vscode-insiders/extensions/ms-vscode.vscode-typescript-next-4.9.20220911/node_modules/typescript/lib/tsserver.js:58736:27)
    at getTypeFromTypeLiteralOrFunctionOrConstructorTypeNode (c:/Users/Daniel/.vscode-insiders/extensions/ms-vscode.vscode-typescript-next-4.9.20220911/node_modules/typescript/lib/tsserver.js:63766:21)
    at getTypeFromTypeNodeWorker (c:/Users/Daniel/.vscode-insiders/extensions/ms-vscode.vscode-typescript-next-4.9.20220911/node_modules/typescript/lib/tsserver.js:64161:28)
    at getTypeFromTypeNode (c:/Users/Daniel/.vscode-insiders/extensions/ms-vscode.vscode-typescript-next-4.9.20220911/node_modules/typescript/lib/tsserver.js:64085:49)
    at getTypeOfNode (c:/Users/Daniel/.vscode-insiders/extensions/ms-vscode.vscode-typescript-next-4.9.20220911/node_modules/typescript/lib/tsserver.js:87993:40)
    at Object.getTypeAtLocation (c:/Users/Daniel/.vscode-insiders/extensions/ms-vscode.vscode-typescript-next-4.9.20220911/node_modules/typescript/lib/tsserver.js:49147:31)
    at Object.getContextualTypeFromParentOrAncestorTypeNode (c:/Users/Daniel/.vscode-insiders/extensions/ms-vscode.vscode-typescript-next-4.9.20220911/node_modules/typescript/lib/tsserver.js:129138:44)
    at c:/Users/Daniel/.vscode-insiders/extensions/ms-vscode.vscode-typescript-next-4.9.20220911/node_modules/typescript/lib/tsserver.js:140822:50
    at Object.mapDefined (c:/Users/Daniel/.vscode-insiders/extensions/ms-vscode.vscode-typescript-next-4.9.20220911/node_modules/typescript/lib/tsserver.js:621:30)
    at c:/Users/Daniel/.vscode-insiders/extensions/ms-vscode.vscode-typescript-next-4.9.20220911/node_modules/typescript/lib/tsserver.js:140819:31
    at Object.flatMap (c:/Users/Daniel/.vscode-insiders/extensions/ms-vscode.vscode-typescript-next-4.9.20220911/node_modules/typescript/lib/tsserver.js:524:25)
    at getReferencesForStringLiteral (c:/Users/Daniel/.vscode-insiders/extensions/ms-vscode.vscode-typescript-next-4.9.20220911/node_modules/typescript/lib/tsserver.js:140817:37)
    at Object.getReferencedSymbolsForNode (c:/Users/Daniel/.vscode-insiders/extensions/ms-vscode.vscode-typescript-next-4.9.20220911/node_modules/typescript/lib/tsserver.js:139547:32)
    at Object.findReferencedSymbols (c:/Users/Daniel/.vscode-insiders/extensions/ms-vscode.vscode-typescript-next-4.9.20220911/node_modules/typescript/lib/tsserver.js:139142:42)
    at Object.findReferences (c:/Users/Daniel/.vscode-insiders/extensions/ms-vscode.vscode-typescript-next-4.9.20220911/node_modules/typescript/lib/tsserver.js:168023:41)
    at c:/Users/Daniel/.vscode-insiders/extensions/ms-vscode.vscode-typescript-next-4.9.20220911/node_modules/typescript/lib/tsserver.js:178178:53
    at searchPosition (c:/Users/Daniel/.vscode-insiders/extensions/ms-vscode.vscode-typescript-next-4.9.20220911/node_modules/typescript/lib/tsserver.js:178375:38)
    at getPerProjectReferences (c:/Users/Daniel/.vscode-insiders/extensions/ms-vscode.vscode-typescript-next-4.9.20220911/node_modules/typescript/lib/tsserver.js:178343:42)
    at getReferencesWorker (c:/Users/Daniel/.vscode-insiders/extensions/ms-vscode.vscode-typescript-next-4.9.20220911/node_modules/typescript/lib/tsserver.js:178175:37)
    at Session.getReferences (c:/Users/Daniel/.vscode-insiders/extensions/ms-vscode.vscode-typescript-next-4.9.20220911/node_modules/typescript/lib/tsserver.js:179778:34)
    at Session.handlers.ts.Map.ts.getEntries._a.<computed> (c:/Users/Daniel/.vscode-insiders/extensions/ms-vscode.vscode-typescript-next-4.9.20220911/node_modules/typescript/lib/tsserver.js:178615:61)
```

Found from https://github.com/mermaid-js/mermaid/blob/3d32280c49125c135cf08a3086e184971a1c6f85/src/utils.ts#L528

### PR DESCRIPTION
**Pull request by a-tarasyuk**
Fixes #50750

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -16498,7 +16498,7 @@ namespace ts {
             if (!links.resolvedType) {
                 // Deferred resolution of members is handled by resolveObjectTypeMembers
                 const aliasSymbol = getAliasSymbolForTypeNode(node);
-                if (getMembersOfSymbol(node.symbol).size === 0 && !aliasSymbol) {
+                if ((node.symbol === undefined || getMembersOfSymbol(node.symbol).size === 0) && !aliasSymbol) {
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts:16501]: How hardened should the checker be against type nodes with no symbol for the language service? Should some other part of the checker have ensured that the JSDoc type node had a symbol? @sandersn @jakebailey

**Review by DanielRosenwasser — APPROVED** [review]

File: src/compiler/checker.ts (hunk 1 above)
Comment (jakebailey) [reply src/compiler/checker.ts:16501]: My uninformed understanding was that this should always be defined (hence, it not being potentially undefined in the Node type); is the binder not walking something?

Comment (a-tarasyuk) [reply src/compiler/checker.ts:16501]: @jakebailey You are right, `binder` is not walking on `jsDoc` nodes in TypeScript files

https://github.com/microsoft/TypeScript/blob/08b91f6b827e9ee98407c0eb969579ab77fa9f0c/src/compiler/binder.ts#L2450-L2464

for this reason, we don't have a `symbol` in the TypeScript files for

https://github.com/microsoft/TypeScript/blob/08b91f6b827e9ee98407c0eb969579ab77fa9f0c/src/compiler/binder.ts#L2637-L2640

At the same time, we can trigger on/use that node in services (in TypeScript/JavaScript)
```ts
// a.ts
/**
 * @type {{ property: /**/"foo"}}
*/
```

in rename 
https://github.com/microsoft/TypeScript/blob/618fb2d8b92357d564ddb8596296ecaba5468002/src/services/rename.ts#L22-L23

or in find all 
https://github.com/microsoft/TypeScript/blob/6362fb2dceb9d4a6fc8640b3d82bca02961d142e/src/services/findAllReferences.ts#L2088-L2089

Comment (jakebailey) [reply src/compiler/checker.ts:16501]: That's what I thought. It feels like if this is the case, then we should have loads of issues, like crashing on hover of a jsdoc in a TS file, for example. But we don't, so does the services code defend against this and we're missing things?

(Also, can we add asserts to the checker to verify that we definitely always have a symbol when given a node?)

Comment (sandersn) [reply src/compiler/checker.ts:16501]: My opinion is that 

1. `@type`, `@typedef` and other type-related jsdoc are meaningless in Typescript and shouldn't be maintained. 
2. That means that find-all-refs, rename, goto-def, et al should intentionally skip jsdoc in Typescript. As @jakebailey points out, we should have loads of crashes already so I think there *must* be some code to skip jsdoc in services already. I think the correct fix is to find that code and make it skip the examples from the bug too.

If there's not already code to skip this kind of jsdoc, and we've just been ignoring a whole segment of crashes, now is *definitely* the time to add it.

**Review by sandersn — CHANGES_REQUESTED** [review]
I'd like to find and augment the skip-typed-jsdoc code in services instead of checking for undefined symbols in the checker.

**Comment by DanielRosenwasser** [comment]
Hm - maybe. What about `@link` which really does need to semantically refer to the TS declaration?

**Comment by sandersn** [comment]
This should only apply to type-related jsdoc. 

`@link` refers to values or types, but it's resolved in a services-only code path iirc. I don't think the binder visits it either. Actually, I don't know how (or whether) rename works on `@link` references in TS or JS.

**Review by sandersn — APPROVED** [review]

File: src/services/utilities.ts [hunk 2]
```diff
@@ -790,6 +790,8 @@ namespace ts {
     }
 
     export function getContextualTypeFromParentOrAncestorTypeNode(node: Expression, checker: TypeChecker): Type | undefined {
+        if (node.flags & (NodeFlags.JSDoc & ~NodeFlags.JavaScriptFile)) return undefined;
```
Comment (DanielRosenwasser) [inline src/services/utilities.ts:793]: Should this have been something like the following?

```ts
if (node.flags & NodeFlags.JSDoc && !(node.flags & NodeFlags.JavaScriptFile)) return undefined;
```

Comment (a-tarasyuk) [reply src/services/utilities.ts:793]: @DanielRosenwasser Yes. Do I need to change it?

Comment (DanielRosenwasser) [reply src/services/utilities.ts:793]: If you wouldn't mind. I think the regression test will look like this

```js
/**
 * @type {"foo" | "bar"}
 */
let x = "foo"$
```

with a find-all-references on `$` resulting in both the expression and the string literal type `"foo"`.

**Pull request merged**